### PR TITLE
Fix/methodhandle polymoprhic dispatch as stated in issue #1108

### PR DESCRIPTION
**Pull request by anddann**
@mbenz89 The current PR just contains test cases to check if the polymorphic methods in `java.lang.invoke.MethodHandle`are treated correctly.

What about the check `jimpleBody.validate()` in the `BafBody` class that was removed in PR https://github.com/Sable/soot/pull/1092?
Due to this missing check this error was somehow "obfuscated" and hidden.

### CODE REVIEW

**Comment by anddann** [comment]
retest this please

File: src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+  @Override
+  protected void runSoot() {
+    PackManager.v().runBodyPacks();
+  }
+
+  @Test
+  public void findsTarget() {
+    final SootMethod sootMethod = prepareTarget("<" + TEST_TARGET_CLASS + ": void test()>", TEST_TARGET_CLASS);
+    Assert.assertNotNull("Could not find target method. System test setup seems to be incorrect.", sootMethod);
```
Comment (mbenz89) [inline src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java]: We might want to place this inside the prepareTarget method directly to inform the user about a potentially missing execution of "mvn test-compil" (or whatever this goal is called) or a potential misspelling of the target method.

File: src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  @Override
+  protected void runSoot() {
+    PackManager.v().runBodyPacks();
+  }
+
+  @Test
+  public void findsTarget() {
+    final SootMethod sootMethod = prepareTarget("<" + TEST_TARGET_CLASS + ": void test()>", TEST_TARGET_CLASS);
```
Comment (mbenz89) [inline src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java]: You might want to use one of the `soot.testing.framework.AbstractTestingFramework#methodSigFromComponents(String, String)` methods to build the target signature.

File: src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  @Override
+  protected void setupSoot() {
+    Options.v().set_allow_phantom_refs(false);
+    Options.v().set_no_bodies_for_excluded(false);
+    Options.v().set_prepend_classpath(true);
+  }
+
+  @Override
+  protected void runSoot() {
+    PackManager.v().runBodyPacks();
```
Comment (mbenz89) [inline src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java]: Not needed anymore due to changes to the framework

File: src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  public void findsTarget() {
+    final SootMethod sootMethod = prepareTarget("<" + TEST_TARGET_CLASS + ": void test()>", TEST_TARGET_CLASS);
+    Assert.assertNotNull("Could not find target method. System test setup seems to be incorrect.", sootMethod);
+    Assert.assertTrue(sootMethod.isConcrete());
+    Assert.assertNotNull(sootMethod.retrieveActiveBody());
+    for (Unit u : sootMethod.getActiveBody().getUnits()) {
+      if (u instanceof AssignStmt) {
+        Value right = ((AssignStmt) u).getRightOp();
+        if (right instanceof InvokeExpr) {
+          // SootMethod m = ((InvokeExpr) right).getMethodRef().resolve();
```
Comment (mbenz89) [inline src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java]: Some actual testing code is needed here^^

File: src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java (hunk 3 above)
Comment (anddann) [reply src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java]: Yes, indeed

File: src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java (hunk 1 above)
Comment (anddann) [reply src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java]: ok

File: src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java (hunk 4 above)
Comment (anddann) [reply src/systemTest/java/soot/jimple/PolymorphicDispatchTest.java]: You are correct

File: src/main/java/soot/Scene.java [hunk 5]
```diff
@@ -1869,6 +1869,10 @@ public void setPkgList(List<String> list) {
   /** Create an unresolved reference to a method. */
   public SootMethodRef makeMethodRef(SootClass declaringClass, String name, List<Type> parameterTypes, Type returnType,
       boolean isStatic) {
+    if (declaringClass.getName().equals(SootMethodRefHandleImpl.METHODHANDLE_SIGNATURE)
```
Comment (mbenz89) [inline src/main/java/soot/Scene.java]: The check may be put into a static method on SootMethodRefHandleImpl. E.g:
`PolymorphicIvokeMethodRef.handlesClass(declaringClass)`.

Ps: Better naming would be preferable...

Comment (mbenz89) [reply src/main/java/soot/Scene.java]: Thereby, if more classes with the SignaturePolymorphic annotation are added later on, there is a single point to add these..

File: src/systemTest/java/soot/testing/framework/AbstractTestingFramework.java [hunk 6]
```diff
@@ -112,6 +113,7 @@ protected SootMethod prepareTarget(String targetMethodSignature, Collection<Stri
     mockStatics();
     SootMethod sootTestMethod = createTestTarget(targetMethodSignature);
     runSoot();
+    Assert.assertNotNull("Could not find target method. System test setup seems to be incorrect.", sootTestMethod);
```
Comment (mbenz89) [inline src/systemTest/java/soot/testing/framework/AbstractTestingFramework.java]: Can you please add something like: "Please try to re-run `mvn test-compile` to make sure that the target code is present for analysis..."

**Comment by mbenz89** [comment]
Is this ready to merge now?

**Comment by anddann** [comment]
Yes, it is.

**Comment by mbenz89** [comment]
Thanks for your contribution!

**Pull request merged**